### PR TITLE
feat(battery_plus): Add Swift Package Manager support

### PR DESCRIPTION
**Pull request by koji-1009**
## Description

Add SPM support.

https://docs.flutter.dev/packages-and-plugins/swift-package-manager/for-plugin-authors

I also noticed an error in the documentation.
Since we need to refer to the documentation when reviewing PRs, it will be draft until the next PR is reviewed.

https://github.com/flutter/website/pull/11020

---

As a working test, I have confirmed that it can be built with flutter 3.24 + CocoaPods, and that it can be built with flutter main channel + SPM.
Note that if you try building with main channel + SPM, you will need to remove some steps from your iOS project, as the script to add Pods is no longer needed.

## Related Issues

- Related #3152

## Checklist

- [x] I read the [Contributor Guide](https://github.com/fluttercommunity/plus_plugins/blob/main/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I titled the PR using [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0).
- [x] I did not modify the `CHANGELOG.md` nor the plugin version in `pubspec.yaml` files.
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate that with a `!` in the title as explained in [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0)).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by koji-1009** [comment]
I would be happy to include flutter/website PR and feel free to review it. There is no need to rush and we would like to discuss how it should be implemented.

**Comment by charafau** [comment]
hi @koji-1009 ! thanks for contribution. I ran pod lint on macos projects and get few warnings:

```
➜  battery_plus pod lib lint macos/battery_plus.podspec  --configuration=Debug --skip-tests --use-modular-headers

 -> battery_plus (0.0.1)
    - WARN  | license: Missing license type.
    - WARN  | [OSX] keys: Missing primary key for `source` attribute. The acceptable ones are: `git, hg, http, svn`.
    - NOTE  | xcodebuild:  note: Using codesigning identity override:
    - NOTE  | [OSX] xcodebuild:  note: Building targets in dependency order
    - NOTE  | [OSX] xcodebuild:  note: Target dependency graph (4 targets)
    - NOTE  | [OSX] xcodebuild:  note: Metadata extraction skipped. No AppIntents.framework dependency found. (in target 'battery_plus' from project 'Pods')
    - NOTE  | [OSX] xcodebuild:  note: Metadata extraction skipped. No AppIntents.framework dependency found. (in target 'App' from project 'App')

[!] battery_plus did not pass validation, due to 2 warnings (but you can use `--allow-warnings` to ignore them).
You can use the `--no-clean` option to inspect any issue.
➜  battery_plus pod lib lint macos/battery_plus.podspec  --configuration=Debug --skip-tests --use-modular-headers --use-libraries


 -> battery_plus (0.0.1)
    - WARN  | license: Missing license type.
    - WARN  | [OSX] keys: Missing primary key for `source` attribute. The acceptable ones are: `git, hg, http, svn`.
    - NOTE  | xcodebuild:  note: Using codesigning identity override:
    - NOTE  | [OSX] xcodebuild:  note: Building targets in dependency order
    - NOTE  | [OSX] xcodebuild:  note: Target dependency graph (4 targets)
    - NOTE  | [OSX] xcodebuild:  note: Metadata extraction skipped. No AppIntents.framework dependency found. (in target 'App' from project 'App')

[!] battery_plus did not pass validation, due to 2 warnings (but you can use `--allow-warnings` to ignore them).
You can use the `--no-clean` option to inspect any issue.
```

not sure if they were there before but it would be nice to fix it

**Review by charafau — COMMENTED** [review]
hi @koji-1009 ! thanks for contribution. I ran pod lint on macos projects and get few warnings:

```
➜  battery_plus pod lib lint macos/battery_plus.podspec  --configuration=Debug --skip-tests --use-modular-headers

 -> battery_plus (0.0.1)
    - WARN  | license: Missing license type.
    - WARN  | [OSX] keys: Missing primary key for `source` attribute. The acceptable ones are: `git, hg, http, svn`.
    - NOTE  | xcodebuild:  note: Using codesigning identity override:
    - NOTE  | [OSX] xcodebuild:  note: Building targets in dependency order
    - NOTE  | [OSX] xcodebuild:  note: Target dependency graph (4 targets)
    - NOTE  | [OSX] xcodebuild:  note: Metadata extraction skipped. No AppIntents.framework dependency found. (in target 'battery_plus' from project 'Pods')
    - NOTE  | [OSX] xcodebuild:  note: Metadata extraction skipped. No AppIntents.framework dependency found. (in target 'App' from project 'App')

[!] battery_plus did not pass validation, due to 2 warnings (but you can use `--allow-warnings` to ignore them).
You can use the `--no-clean` option to inspect any issue.
➜  battery_plus pod lib lint macos/battery_plus.podspec  --configuration=Debug --skip-tests --use-modular-headers --use-libraries


 -> battery_plus (0.0.1)
    - WARN  | license: Missing license type.
    - WARN  | [OSX] keys: Missing primary key for `source` attribute. The acceptable ones are: `git, hg, http, svn`.
    - NOTE  | xcodebuild:  note: Using codesigning identity override:
    - NOTE  | [OSX] xcodebuild:  note: Building targets in dependency order
    - NOTE  | [OSX] xcodebuild:  note: Target dependency graph (4 targets)
    - NOTE  | [OSX] xcodebuild:  note: Metadata extraction skipped. No AppIntents.framework dependency found. (in target 'App' from project 'App')

[!] battery_plus did not pass validation, due to 2 warnings (but you can use `--allow-warnings` to ignore them).
You can use the `--no-clean` option to inspect any issue.
```

not sure if they were there before but it would be nice to fix it

**Review by charafau — APPROVED** [review]
sorry, I think that warning were there before, so let's leave it as it is. Probably fix for different PR.

lgtm

**Comment by koji-1009** [comment]
The contents of the `ios/` directory will not be uploaded to https://cocoapods.org/. So `pod spec lint` shows multiple problems, but I don't see any major issues.
I'm sure time will solve the problem in the future once the migration from CocoaPods to SPM is complete.

**Review by miquelbeltran — COMMENTED** [review]
LGTM, I still want to test this on my devices before merge.

File: packages/battery_plus/battery_plus/macos/battery_plus/Package.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    ],
+    dependencies: [],
+    targets: [
+        .target(
+            name: "battery_plus",
+            dependencies: [],
+            resources: []
+        )
+    ]
+)
```
Comment (loic-sharma) [inline packages/battery_plus/battery_plus/macos/battery_plus/Package.swift:22]: Nit: add new line here

File: packages/battery_plus/battery_plus/ios/battery_plus.podspec [hunk 2]
```diff
@@ -14,11 +14,11 @@ Downloaded by pub (not CocoaPods).
...
-  s.source_files = 'Classes/**/*'
-  s.public_header_files = 'Classes/**/*.h'
+  s.source_files = 'battery_plus/Sources/battery_plus/**/*.{h,m}'
+  s.public_header_files = 'battery_plus/Sources/battery_plus/include/**/*.h'
   s.dependency 'Flutter'
 
   s.platform = :ios, '12.0'
   s.pod_target_xcconfig = { 'DEFINES_MODULE' => 'YES' }
-  s.resource_bundles = {'batery_plus_privacy' => ['PrivacyInfo.xcprivacy']}
+  s.resource_bundles = {'batery_plus_privacy' => ['battery_plus/Sources/battery_plus/PrivacyInfo.xcprivacy']}
```
Comment (loic-sharma) [inline packages/battery_plus/battery_plus/ios/battery_plus.podspec:23]: Should `batery` be `battery` here?

**Review by vbuberen — APPROVED** [review]
Did some testing on iOS and MacOS and can confirm that everything works as it should in both SPM enabled and disabled modes.

Thanks a lot for your contribution. We are good to merge this PR!

**Pull request merged**